R/rdeephaven/src/client.cpp

```diff
@@ -0,0 +1,233 @@
+#include <iostream>
+#include <utility>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "deephaven/client/client.h"
+#include "deephaven/client/flight.h"
+
+#include <arrow/c/abi.h>
+#include <arrow/c/bridge.h>
+
+#include <Rcpp.h>
+
+// forward declaration of classes
+class ClientOptionsWrapper;
+class ClientWrapper;
+
+// ######################### DH WRAPPERS #########################
+
+class TableHandleWrapper {
+public:
+    TableHandleWrapper(deephaven::client::TableHandle ref_table) : internal_tbl_hdl(std::move(ref_table)) {};
+
+    // DEEPHAVEN QUERY METHODS WILL GO HERE
+
+    /**
+     * Binds the table referenced by this table handle to a variable on the server called tableName.
+     * Without this call, new tables are not accessible from the client.
+     * @param tableName Name for the new table on the server.
+    */
+    void bindToVariable(std::string tableName) {
+        internal_tbl_hdl.bindToVariable(tableName);
+    }
+
+    /**
+     * Creates and returns a pointer to an ArrowArrayStream C struct containing the data from the table referenced by internal_tbl_hdl.
+     * Intended to be used for creating an Arrow RecordBatchReader in R via RecordBatchReader$import_from_c(ptr).
+    */
+    SEXP getArrowArrayStreamPtr() {
+
+        std::shared_ptr<arrow::flight::FlightStreamReader> fsr = internal_tbl_hdl.getFlightStreamReader();
+
+        std::vector<std::shared_ptr<arrow::RecordBatch>> empty_record_batches;
+        DEEPHAVEN_EXPR_MSG(fsr->ReadAll(&empty_record_batches)); // need to add OK or throw
+
+        std::shared_ptr<arrow::RecordBatchReader> record_batch_reader = arrow::RecordBatchReader::Make(empty_record_batches).ValueOrDie();
+        ArrowArrayStream* stream_ptr = new ArrowArrayStream();
+        arrow::ExportRecordBatchReader(record_batch_reader, stream_ptr);
+
+        // XPtr is needed here to ensure Rcpp can properly handle type casting, as it does not like raw pointers
+        return Rcpp::XPtr<ArrowArrayStream>(stream_ptr, true);
+    };
+
+private:
+    deephaven::client::TableHandle internal_tbl_hdl;
+};
+
+
+
+class ClientOptionsWrapper {
+public:
+
+    ClientOptionsWrapper() {
+        internal_options = new deephaven::client::ClientOptions();
+        bool authentication_set = false;
+        bool session_set = false;
```

Comment:
Is there a reason this isn't just left to the non-wrapper code?

Reply:
Nope. I was going to use them to check if ClientOptions had already been set, but I should do that in R anyway.

Reply:
Why can't you just leave it to C++?

Reply:
I thought better to catch earlier if possible?